File: back/src/test/java/com/tdevilleduc/urthehero/back/service/PersonServiceTest.java
package com.tdevilleduc.urthehero.back.service;

import com.tdevilleduc.urthehero.back.AbstractTest;
import com.tdevilleduc.urthehero.back.BackApplication;
import com.tdevilleduc.urthehero.back.exceptions.PersonNotFoundException;
import com.tdevilleduc.urthehero.back.model.Person;
import com.tdevilleduc.urthehero.back.service.impl.PersonService;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.junit.jupiter.SpringExtension;

import java.util.Optional;

import static org.junit.jupiter.api.Assertions.assertEquals;

@ExtendWith(SpringExtension.class)
@SpringBootTest(classes = BackApplication.class)
class PersonServiceTest extends AbstractTest {

    @Autowired
    private PersonService personService;

    @Test
    void test_exists_thenCorrect() {
        Integer personId = 1;
        boolean exists = personService.exists(personId);
        Assertions.assertTrue(exists);
    }

    @Test
    void test_notExists_thenCorrect() {
        Integer personId = 41;
        boolean notExists = personService.notExists(personId);
        Assertions.assertTrue(notExists);
    }

    @Test
    void test_findById_thenCorrect() {
        Integer personId = 1;
        Optional<Person> optional = personService.findById(personId);
        Assertions.assertTrue(optional.isPresent());
        Person person = optional.get();

        Assertions.assertNotNull(person);
        Assertions.assertEquals(Integer.valueOf(1), person.getId());
        Assertions.assertEquals("Thomas Deville-Duc", person.getDisplayName());
        Assertions.assertEquals("thomas@gmail.com", person.getEmail());
        Assertions.assertEquals("tdevilleduc", person.getLogin());
        Assertions.assertEquals("password", person.getPassword());
    }

    @Test
    void test_findById_thenNotFound() {
        Integer personId = 13;
        Optional<Person> optional = personService.findById(personId);
        Assertions.assertTrue(optional.isEmpty());
    }

    @Test
    void findByIdWithIdNull() {
        // with resilience4j, IllegalArgumentException is catched and fallback method 'emptyPerson' is called
//        Assertions.assertThrows(IllegalArgumentException.class, () -> personService.findById(null));
        assertEquals(Optional.empty(), personService.findById(null));
    }

    @Test
    void deleteWithIdNull() {
        Assertions.assertThrows(IllegalArgumentException.class, () -> personService.delete(null));
    }

    @Test
    void delete_thenNotFound() {
        Integer personId = 13;
        Assertions.assertThrows(PersonNotFoundException.class, () -> personService.delete(personId));
    }
}

File: back/src/test/java/com/tdevilleduc/urthehero/back/service/PersonServiceTest.kt
package com.tdevilleduc.urthehero.back.service

import com.tdevilleduc.urthehero.back.dao.UserDao
import com.tdevilleduc.urthehero.back.exceptions.UserNotFoundException
import com.tdevilleduc.urthehero.back.service.impl.UserService
import com.tdevilleduc.urthehero.back.util.TestUtil
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.jupiter.MockitoExtension
import org.testcontainers.shaded.org.apache.commons.lang.RandomStringUtils
import java.util.*

@ExtendWith(MockitoExtension::class)
internal class PersonServiceTest {
    private val random: Random = Random()

    @InjectMocks
    private lateinit var userService: UserService
    @Mock
    private lateinit var userDao: UserDao

    @Test
    fun test_exists_withIdNull() {
        val userId = null
        val exists = userService.exists(userId)
        Assertions.assertFalse(exists)
    }

    @Test
    fun test_notExists_thenSuccess() {
        val userId = random.nextInt()
        Mockito.`when`(userDao.findById(userId)).thenReturn(Optional.empty())

        val notExists = userService.notExists(userId)
        Assertions.assertTrue(notExists)
    }

    @Test
    fun test_findById_thenCorrect() {
        val expectedUser = TestUtil.createUser()
        val userId = random.nextInt()
        expectedUser.userId = userId
        Mockito.`when`(userDao.findById(expectedUser.userId!!)).thenReturn(Optional.of(expectedUser))

        val user = userService.findById(expectedUser.userId!!)
        Assertions.assertNotNull(user)
        Assertions.assertEquals(userId, user.userId)
        Assertions.assertEquals(expectedUser.username, user.username)
        Assertions.assertEquals(expectedUser.password, user.password)
    }
    @Test
    fun test_findById_thenNotFound() {
        val userId = random.nextInt()
        Mockito.`when`(userDao.findById(userId)).thenReturn(Optional.empty())

        Assertions.assertThrows(UserNotFoundException::class.java) { userService.findById(userId) }
    }

    @Test
    fun delete_thenNotFound() {
        val userId = random.nextInt()
        Mockito.`when`(userDao.findById(userId)).thenReturn(Optional.empty())

        Assertions.assertThrows(UserNotFoundException::class.java) { userService.delete(userId) }
    }

    @Test
    fun test_findByUsername_thenCorrect() {
        val expectedUser = TestUtil.createUser()
        Mockito.`when`(userDao.findByUsername(expectedUser.username)).thenReturn(Optional.of(expectedUser))

        val user = userService.findByUsername(expectedUser.username)
        Assertions.assertNotNull(user)
        Assertions.assertEquals(expectedUser.userId, user.userId)
        Assertions.assertEquals(expectedUser.username, user.username)
        Assertions.assertEquals(expectedUser.password, user.password)
    }

    @Test
    fun test_findByUsername_thenNotFound() {
        val userUsername = RandomStringUtils.random(20)
        Mockito.`when`(userDao.findByUsername(userUsername)).thenReturn(Optional.empty())

        Assertions.assertThrows(UserNotFoundException::class.java) {
            userService.findByUsername(userUsername)
        }
    }
}